In [409]:
import pandas as pd 
import numpy as np
import sqlite3
import matplotlib.pyplot as plt
import seaborn as sns


### Setting up each file as a DataFrame and checking that they can be read. 

In [410]:
bech_df = pd.read_csv("../Data/Bechdel_IMDB_Merge0524.csv")
cin_df = pd.read_csv("../Data/cinatomy_movie_profiles.csv")
osc_df = pd.read_csv("../Data/oscars_df.csv")
horror_df = pd.read_csv("../Data/horror_movies.csv")


In [411]:
bech_df.head()

,title,year,imdbid,id,bechdelRating,imdbAverageRating,numVotes,runtimeMinutes,genre1,genre2,genre3
0,Miss Jerry,1894,9.0,9779,0,5.4,212.0,45,Romance,NaN,NaN
1,"Story of the Kelly Gang, The",1906,574.0,1349,1,6.0,903.0,70,Action,Adventure,Biography
2,Cleopatra,1912,2101.0,2003,2,5.1,622.0,100,Drama,History,NaN
3,A Florida Enchantment,1914,3973.0,4457,2,5.8,300.0,63,Comedy,NaN,NaN
4,"Birth of a Nation, The",1915,4972.0,1258,2,6.1,26403.0,195,Drama,History,War


In [412]:
cin_df.head()

,movie_title,pacing_efficiency,originality,family_friendly,plot_quality_rating,plot_quality_reasons_for_liking,plot_quality_reasons_for_disliking,dialogues,end_feeling_emotional_adjective,end_feeling_ending_rating,...,sound_bgm_quality,sound_song_tracks_quality,performance_of_actors,connected,references_contained,cognitive_requirement,technical_knowledge_required,photosensitivity_warnings,animal_harm,trailer_or_spoiler
0,The Shawshank Redemption,4,4,2,5,"['unpredictable', 'plot_twists', 'strong_pacin...",[],5,happy,5,...,5,5,5,5,1,3,1,1,1,1
1,The Godfather,4,5,1,5,"['strong_pacing', 'tight_structure']",[],5,bittersweet,5,...,5,4,5,3,4,4,1,1,4,3
2,The Dark Knight,5,5,1,5,"['unpredictable', 'plot_twists', 'strong_pacin...",[],5,bittersweet,5,...,5,4,5,5,4,4,1,3,1,5
3,The Godfather: Part II,4,5,1,5,"['unpredictable', 'plot_twists', 'strong_pacin...",[],5,bittersweet,5,...,5,4,5,5,5,4,1,1,1,5
4,12 Angry Men,5,5,3,5,"['unpredictable', 'plot_twists', 'strong_pacin...",[],5,energized,5,...,3,1,5,1,1,4,1,1,1,3


In [413]:
osc_df.head()

,Unnamed: 0,Film,Oscar Year,Film Studio/Producer(s),Award,Year of Release,Movie Time,Movie Genre,IMDB Rating,IMDB Votes,...,Tomatometer Status,Tomatometer Rating,Tomatometer Count,Audience Status,Audience Rating,Audience Count,Tomatometer Top Critics Count,Tomatometer Fresh Critics Count,Tomatometer Rotten Critics Count,Film ID
0,0,Wings,1927/28,Famous Players-Lasky,Winner,1927,144,"Drama,Romance,War",7.5,"12,221",...,Certified-Fresh,93.0,46.0,Upright,78.0,3530.0,9.0,43.0,3.0,2becf7d5-a3de-46ab-ae45-abdd6b588067
1,1,7th Heaven,1927/28,Fox,Nominee,1927,110,"Drama,Romance",7.7,"3,439",...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,19ed3295-a878-4fd2-8e60-5cd7b5f93dad
2,2,The Racket,1927/28,The Caddo Company,Nominee,1928,84,"Crime,Drama,Film-Noir",6.7,"1,257",...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3111c2d8-0908-4093-8ff3-99c89f2f2f08
3,3,The Broadway Melody,1928/29,Metro-Goldwyn-Mayer,Winner,1929,100,"Drama,Musical,Romance",5.7,"6,890",...,Rotten,33.0,24.0,Spilled,21.0,1813.0,7.0,8.0,16.0,de063f3f-2d35-4e1c-8636-6eb4c16bd236
4,4,Alibi,1928/29,Feature Productions,Nominee,1929,91,"Action,Crime,Romance",5.8,765,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,609887c2-877c-43a4-b88c-e40e31096a98


In [414]:
horror_df.head()

,id,original_title,title,original_language,overview,tagline,release_date,poster_path,popularity,vote_count,vote_average,budget,revenue,runtime,status,adult,backdrop_path,genre_names,collection,collection_name
0,760161,Orphan: First Kill,Orphan: First Kill,en,After escaping from an Estonian psychiatric fa...,There's always been something wrong with Esther.,2022-07-27,/pHkKbIRoCe7zIFvqan9LFSaQAde.jpg,5088.584,902,6.9,0,9572765,99,Released,False,/5GA3vV1aWWHTSDO5eno8V5zDo8r.jpg,"Horror, Thriller",760193.0,Orphan Collection
1,760741,Beast,Beast,en,A recently widowed man and his two teenage dau...,Fight for family.,2022-08-11,/xIGr7UHsKf0URWmyyd5qFMAq4d8.jpg,2172.338,584,7.1,0,56000000,93,Released,False,/2k9tBql5GYH328Krj66tDT9LtFZ.jpg,"Adventure, Drama, Horror",NaN,NaN
2,882598,Smile,Smile,en,"After witnessing a bizarre, traumatic incident...","Once you see it, it’s too late.",2022-09-23,/hiaeZKzwsk4y4atFhmncO5KRxeT.jpg,1863.628,114,6.8,17000000,45000000,115,Released,False,/mVNPfpydornVe4H4UCIk7WevWjf.jpg,"Horror, Mystery, Thriller",NaN,NaN
3,756999,The Black Phone,The Black Phone,en,"Finney Blake, a shy but clever 13-year-old boy...",Never talk to strangers.,2022-06-22,/lr11mCT85T1JanlgjMuhs9nMht4.jpg,1071.398,2736,7.9,18800000,161000000,103,Released,False,/AfvIjhDu9p64jKcmohS4hsPG95Q.jpg,"Horror, Thriller",NaN,NaN
4,772450,Presencias,Presences,es,A man who loses his wife and goes to seclude h...,NaN,2022-09-07,/dgDT3uol3mdvwEg0jt1ble3l9hw.jpg,1020.995,83,7.0,0,0,0,Released,False,/ojfzhdwRemcDt1I6pao6vVLw9AA.jpg,Horror,NaN,NaN


### Checking each dataset's info, shape, and if there are any duplicate rows.

In [415]:
bech_df.info()
bech_df.duplicated().value_counts()

<class 'pandas.DataFrame'>
RangeIndex: 9718 entries, 0 to 9717
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   title              9718 non-null   str    
 1   year               9718 non-null   int64  
 2   imdbid             9718 non-null   float64
 3   id                 9718 non-null   int64  
 4   bechdelRating      9718 non-null   int64  
 5   imdbAverageRating  9718 non-null   float64
 6   numVotes           9718 non-null   float64
 7   runtimeMinutes     9718 non-null   str    
 8   genre1             9718 non-null   str    
 9   genre2             8446 non-null   str    
 10  genre3             5807 non-null   str    
dtypes: float64(3), int64(3), str(5)
memory usage: 835.3 KB


False    9718
Name: count, dtype: int64

In [416]:
cin_df.info()
cin_df.duplicated().value_counts()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 27 columns):
 #   Column                              Non-Null Count  Dtype
---  ------                              --------------  -----
 0   movie_title                         1000 non-null   str  
 1   pacing_efficiency                   1000 non-null   int64
 2   originality                         1000 non-null   int64
 3   family_friendly                     1000 non-null   int64
 4   plot_quality_rating                 1000 non-null   int64
 5   plot_quality_reasons_for_liking     1000 non-null   str  
 6   plot_quality_reasons_for_disliking  1000 non-null   str  
 7   dialogues                           1000 non-null   int64
 8   end_feeling_emotional_adjective     1000 non-null   str  
 9   end_feeling_ending_rating           1000 non-null   int64
 10  end_feeling_structural_adjective    1000 non-null   str  
 11  immersive                           1000 non-null   int64
 12  impactful         

False    999
True       1
Name: count, dtype: int64

In [417]:
osc_df.info()
osc_df.duplicated().value_counts()


<class 'pandas.DataFrame'>
RangeIndex: 571 entries, 0 to 570
Data columns (total 30 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Unnamed: 0                        571 non-null    int64  
 1   Film                              571 non-null    str    
 2   Oscar Year                        571 non-null    str    
 3   Film Studio/Producer(s)           571 non-null    str    
 4   Award                             571 non-null    str    
 5   Year of Release                   571 non-null    int64  
 6   Movie Time                        571 non-null    int64  
 7   Movie Genre                       571 non-null    str    
 8   IMDB Rating                       571 non-null    float64
 9   IMDB Votes                        571 non-null    str    
 10  Movie Info                        438 non-null    str    
 11  Genres                            439 non-null    str    
 12  Critic Consensus   

False    571
Name: count, dtype: int64

In [418]:
horror_df.info()
horror_df.duplicated().value_counts()

<class 'pandas.DataFrame'>
RangeIndex: 32540 entries, 0 to 32539
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 32540 non-null  int64  
 1   original_title     32540 non-null  str    
 2   title              32540 non-null  str    
 3   original_language  32540 non-null  str    
 4   overview           31254 non-null  str    
 5   tagline            12706 non-null  str    
 6   release_date       32540 non-null  str    
 7   poster_path        28066 non-null  str    
 8   popularity         32540 non-null  float64
 9   vote_count         32540 non-null  int64  
 10  vote_average       32540 non-null  float64
 11  budget             32540 non-null  int64  
 12  revenue            32540 non-null  int64  
 13  runtime            32540 non-null  int64  
 14  status             32540 non-null  str    
 15  adult              32540 non-null  bool   
 16  backdrop_path      13545 non-null

False    32540
Name: count, dtype: int64

In [419]:
print(bech_df.shape)
print(cin_df.shape)
print(osc_df.shape)
print(horror_df.shape)

(9718, 11)
(1000, 27)
(571, 30)
(32540, 20)


Dropping duplicates 

In [420]:
cin_df.drop_duplicates(keep='first', inplace=True)
cin_df.duplicated().value_counts()

False    999
Name: count, dtype: int64

#### Checking for nulls, and determining if the columns need adjusting or rows need dropping

In [421]:
bech_df.isnull().sum()

title                   0
year                    0
imdbid                  0
id                      0
bechdelRating           0
imdbAverageRating       0
numVotes                0
runtimeMinutes          0
genre1                  0
genre2               1272
genre3               3911
dtype: int64

Only nulls in the Bechdel Test Dataframe are from subgenres(columns: genre2, genre3)
Most likely needs no adjusting. 

In [422]:
cin_df.isnull().sum()

movie_title                           0
pacing_efficiency                     0
originality                           0
family_friendly                       0
plot_quality_rating                   0
plot_quality_reasons_for_liking       0
plot_quality_reasons_for_disliking    0
dialogues                             0
end_feeling_emotional_adjective       0
end_feeling_ending_rating             0
end_feeling_structural_adjective      0
immersive                             0
impactful                             0
overall_feeling                       0
visuals                               0
visual_effects                        0
sound_audio_balance_issue             0
sound_bgm_quality                     0
sound_song_tracks_quality             0
performance_of_actors                 0
connected                             0
references_contained                  0
cognitive_requirement                 0
technical_knowledge_required          0
photosensitivity_warnings             0


No nulls found in Cinatomy DataFrame

In [423]:
osc_df.isnull().sum()

Unnamed: 0                            0
Film                                  0
Oscar Year                            0
Film Studio/Producer(s)               0
Award                                 0
Year of Release                       0
Movie Time                            0
Movie Genre                           0
IMDB Rating                           0
IMDB Votes                            0
Movie Info                          133
Genres                              132
Critic Consensus                    232
Content Rating                      132
Directors                           134
Authors                             134
Actors                              132
Original Release Date               132
Streaming Release Date              133
Production Company                  132
Tomatometer Status                  132
Tomatometer Rating                  132
Tomatometer Count                   132
Audience Status                     135
Audience Rating                     132


Presumably, 132 rows are missing information in the following columns:
- movie info*
- genres*
- critic consensus
- content rating
- directors
- authors
- actors
- original release date
- streaming release date
- production company
- ... and all columns involving Rotten Tomatoes

*( ***asterick*** where null in column might be a concern moving forward)*


In [424]:
print("Most Recent Release Date:")
print(osc_df['Original Release Date'].max())
print("Earliest Release Date:")
print(osc_df['Original Release Date'].min())


Most Recent Release Date:
2019-11-06
Earliest Release Date:
1927-08-12


The cause of missing values might be the earlier movies. 
***Look more into this in later cleaning.***

In [425]:
horror_df.isnull().sum()

id                       0
original_title           0
title                    0
original_language        0
overview              1286
tagline              19834
release_date             0
poster_path           4474
popularity               0
vote_count               0
vote_average             0
budget                   0
revenue                  0
runtime                  0
status                   0
adult                    0
backdrop_path        18995
genre_names              0
collection           30234
collection_name      30234
dtype: int64

Nulls in collection_name are fine as is, they are there to let us know that that particular movie is not part of a larger collection.

Nulls in tagline are also fine, not all movies have a tagline, especially the older movies in this dataframe. 
Might ultimately drop this column as it's not incredibly useful. 
Though it could be interesting to see why there is a lack of tagline? 

Columns to drop(due to nulls or repetition): 
- collection (unnecessary since we have collection_name)
- adult(none of these movies are pornography, so all rows have same value)
I don't plan to use these columns and they have a significt number of nulls.


### Dropping unnecessary columns
cin_df is very clean already, so I will not be dropping any columns from this dataset.  

From bech_df i'll be dropping the id column only.
  
From osc_df I will be dropping several columns thaat i do not plan to use for this project:
IMDB Votes, Streaming Release Date, Tomatometer Count, Audience Status, Audience Rating, Audience Count, Tomatometer Top Critics Count, Tomatometer Fresh Critics Count, Tomatometer Rotten Critics Count, Film ID, Unnamed
  
From horror_df I will be dropping poster_path and backdrop_path(as well as adult and collection columns) since I will not be using them in this project.  



In [426]:
horror_df.drop(['adult', 'id','collection', 'poster_path','backdrop_path', 'vote_count', 'vote_average'], axis=1, inplace=True)
#check if columns dropped
horror_df.info()


<class 'pandas.DataFrame'>
RangeIndex: 32540 entries, 0 to 32539
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   original_title     32540 non-null  str    
 1   title              32540 non-null  str    
 2   original_language  32540 non-null  str    
 3   overview           31254 non-null  str    
 4   tagline            12706 non-null  str    
 5   release_date       32540 non-null  str    
 6   popularity         32540 non-null  float64
 7   budget             32540 non-null  int64  
 8   revenue            32540 non-null  int64  
 9   runtime            32540 non-null  int64  
 10  status             32540 non-null  str    
 11  genre_names        32540 non-null  str    
 12  collection_name    2306 non-null   str    
dtypes: float64(1), int64(3), str(9)
memory usage: 3.2 MB


In [427]:
osc_df.drop(['Unnamed: 0','IMDB Votes', 'Streaming Release Date', 'Actors','Year of Release', 'Tomatometer Count', 'Audience Status', 'Audience Rating', 'Audience Count', 'Tomatometer Top Critics Count', 'Tomatometer Fresh Critics Count', 'Tomatometer Rotten Critics Count', 'Film ID'], axis=1, inplace=True)
#check if columns dropped
osc_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 571 entries, 0 to 570
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Film                     571 non-null    str    
 1   Oscar Year               571 non-null    str    
 2   Film Studio/Producer(s)  571 non-null    str    
 3   Award                    571 non-null    str    
 4   Movie Time               571 non-null    int64  
 5   Movie Genre              571 non-null    str    
 6   IMDB Rating              571 non-null    float64
 7   Movie Info               438 non-null    str    
 8   Genres                   439 non-null    str    
 9   Critic Consensus         339 non-null    str    
 10  Content Rating           439 non-null    str    
 11  Directors                437 non-null    str    
 12  Authors                  437 non-null    str    
 13  Original Release Date    439 non-null    str    
 14  Production Company       439 non-null

In [428]:
bech_df.drop(["id"], axis=1, inplace=True)
#check if column dropped
bech_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9718 entries, 0 to 9717
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   title              9718 non-null   str    
 1   year               9718 non-null   int64  
 2   imdbid             9718 non-null   float64
 3   bechdelRating      9718 non-null   int64  
 4   imdbAverageRating  9718 non-null   float64
 5   numVotes           9718 non-null   float64
 6   runtimeMinutes     9718 non-null   str    
 7   genre1             9718 non-null   str    
 8   genre2             8446 non-null   str    
 9   genre3             5807 non-null   str    
dtypes: float64(3), int64(2), str(5)
memory usage: 759.3 KB


### Dropping rows  
Dropping rows from horror_df that have nulls in the overview, the number of rows being dropped is a small number in comparison to the dataset as a whole. 




In [429]:
horror_df.dropna(subset=['overview'], inplace=True)
#check if rows dropped
horror_df.isnull().sum()

original_title           0
title                    0
original_language        0
overview                 0
tagline              18579
release_date             0
popularity               0
budget                   0
revenue                  0
runtime                  0
status                   0
genre_names              0
collection_name      28970
dtype: int64

In [430]:
# drop rows where status not equal to released 

un_released = horror_df[horror_df['status'] != 'Released'].index
horror_df.drop(un_released, inplace=True)

#check rows were droppecd
horror_df['status'].value_counts()


status
Released    31142
Name: count, dtype: int64

As suspected, movies with no movie info are earlier movies(before 1958), based of this I believe it'll be fine drop these rows.

In [431]:
no_movieinfo = osc_df[osc_df['Movie Info'].isnull()] 
no_movieinfo['Original Release Date'].max()

'1958-09-20'

Dropping all rows with null in the Movie Info column also resolved the nulls in almost every other column (significantly- Genres column)

In [432]:
osc_df.dropna(subset=['Movie Info'], inplace=True)
#check if rows dropped, how other columns were effected as well
osc_df.isnull().sum()

Film                         0
Oscar Year                   0
Film Studio/Producer(s)      0
Award                        0
Movie Time                   0
Movie Genre                  0
IMDB Rating                  0
Movie Info                   0
Genres                       0
Critic Consensus           100
Content Rating               0
Directors                    2
Authors                      2
Original Release Date        0
Production Company           0
Tomatometer Status           0
Tomatometer Rating           0
dtype: int64

### Replace Values

In [433]:
#replacing 0 runtime values with NaN to avoid skewing data
#researched a few movies with '0' runtimes, obviously not correct.

horror_df['runtime'] = horror_df['runtime'].replace(0, np.nan)
horror_df['runtime'].isnull().sum()

np.int64(1652)

In [434]:
bech_df['runtimeMinutes'] = bech_df['runtimeMinutes'].replace("\\N", np.nan)
#when trying to change the datatype of runtimeMinutes column, I was getting a ValueError: invalid literal for int() with base 10: '\N' error.

datatype_issues = bech_df[bech_df['runtimeMinutes'] == "\\N"] 
datatype_issues['runtimeMinutes'].value_counts()


Series([], Name: count, dtype: int64)

### Fix datatypes  
  
Checking the datatypes are appropriate and changing them if needed.  
Changing release date/year of horror_df, bech_df, and osc_df to datetime.  
Oscar Year format is Year/Year ex. 1986/87.  
**need to figure out what to do about oscar year format, could possibly keep as string?**




In [435]:

horror_df['release_date'] = pd.to_datetime(horror_df['release_date'])
bech_df['year'] = pd.to_datetime(bech_df['year'])
osc_df['Original Release Date'] = pd.to_datetime(osc_df['Original Release Date'])



In [436]:
bech_df['runtimeMinutes'] = bech_df['runtimeMinutes'].astype(float)
osc_df['Movie Time'] = osc_df['Movie Time'].astype(float)


In [437]:
horror_df.dtypes

original_title                  str
title                           str
original_language               str
overview                        str
tagline                         str
release_date         datetime64[us]
popularity                  float64
budget                        int64
revenue                       int64
runtime                     float64
status                          str
genre_names                     str
collection_name                 str
dtype: object

In [438]:
bech_df.dtypes

title                           str
year                 datetime64[ns]
imdbid                      float64
bechdelRating                 int64
imdbAverageRating           float64
numVotes                    float64
runtimeMinutes              float64
genre1                          str
genre2                          str
genre3                          str
dtype: object

In [439]:
cin_df.dtypes

movie_title                             str
pacing_efficiency                     int64
originality                           int64
family_friendly                       int64
plot_quality_rating                   int64
plot_quality_reasons_for_liking         str
plot_quality_reasons_for_disliking      str
dialogues                             int64
end_feeling_emotional_adjective         str
end_feeling_ending_rating             int64
end_feeling_structural_adjective        str
immersive                             int64
impactful                             int64
overall_feeling                       int64
visuals                               int64
visual_effects                        int64
sound_audio_balance_issue              bool
sound_bgm_quality                     int64
sound_song_tracks_quality             int64
performance_of_actors                 int64
connected                             int64
references_contained                  int64
cognitive_requirement           

In [440]:
osc_df.dtypes

Film                                  str
Oscar Year                            str
Film Studio/Producer(s)               str
Award                                 str
Movie Time                        float64
Movie Genre                           str
IMDB Rating                       float64
Movie Info                            str
Genres                                str
Critic Consensus                      str
Content Rating                        str
Directors                             str
Authors                               str
Original Release Date      datetime64[us]
Production Company                    str
Tomatometer Status                    str
Tomatometer Rating                float64
dtype: object

### Rename Columns 
For consistency, when we start making SQL tables and queries.

In [441]:
horror_df.rename(columns={'popularity': 'tmdb_rating', 'genre_names': 'movie_genres'}, inplace=True)

#checking if column name changed 
horror_df.info()

<class 'pandas.DataFrame'>
Index: 31142 entries, 0 to 32537
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   original_title     31142 non-null  str           
 1   title              31142 non-null  str           
 2   original_language  31142 non-null  str           
 3   overview           31142 non-null  str           
 4   tagline            12628 non-null  str           
 5   release_date       31142 non-null  datetime64[us]
 6   tmdb_rating        31142 non-null  float64       
 7   budget             31142 non-null  int64         
 8   revenue            31142 non-null  int64         
 9   runtime            29490 non-null  float64       
 10  status             31142 non-null  str           
 11  movie_genres       31142 non-null  str           
 12  collection_name    2281 non-null   str           
dtypes: datetime64[us](1), float64(2), int64(2), str(8)
memory usage: 3.3 MB


In [442]:
bech_df.rename(columns={
    "year": "release_date",
    "imdbAverageRating": "imdb_rating",
    "bechdelRating": "bechdel_rating",
    "runtimeMinutes": "runtime",
    "numVotes": "imdb_votes"
    }, inplace=True)

#checking if column names changed
bech_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9718 entries, 0 to 9717
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   title           9718 non-null   str           
 1   release_date    9718 non-null   datetime64[ns]
 2   imdbid          9718 non-null   float64       
 3   bechdel_rating  9718 non-null   int64         
 4   imdb_rating     9718 non-null   float64       
 5   imdb_votes      9718 non-null   float64       
 6   runtime         9710 non-null   float64       
 7   genre1          9718 non-null   str           
 8   genre2          8446 non-null   str           
 9   genre3          5807 non-null   str           
dtypes: datetime64[ns](1), float64(4), int64(1), str(4)
memory usage: 759.3 KB


In [443]:
cin_df.rename(columns={"movie_title": "title"}, inplace=True)
#checking if column name changed
cin_df.info()

<class 'pandas.DataFrame'>
Index: 999 entries, 0 to 999
Data columns (total 27 columns):
 #   Column                              Non-Null Count  Dtype
---  ------                              --------------  -----
 0   title                               999 non-null    str  
 1   pacing_efficiency                   999 non-null    int64
 2   originality                         999 non-null    int64
 3   family_friendly                     999 non-null    int64
 4   plot_quality_rating                 999 non-null    int64
 5   plot_quality_reasons_for_liking     999 non-null    str  
 6   plot_quality_reasons_for_disliking  999 non-null    str  
 7   dialogues                           999 non-null    int64
 8   end_feeling_emotional_adjective     999 non-null    str  
 9   end_feeling_ending_rating           999 non-null    int64
 10  end_feeling_structural_adjective    999 non-null    str  
 11  immersive                           999 non-null    int64
 12  impactful               

In [444]:
osc_df.rename(columns={
    "Film": "title",
    "Oscar Year": "oscar_year",
    "Film Studio/Producer(s)": "studio_producer",
    "Award": "oscar_award",
    "Movie Time": "runtime",
    "Movie Info": "overview",
    "Genres": "genres",
    "Critic Consensus": "critic_consensus",
    "Content Rating": "content_rating",
    "Directors": "directors",
    "Authors": "authors",
    "Original Release Date": "release_date",
    "Movie Genre": "movie_genre",
    "Production Company": "production_company",
    "Tomatometer Rating": "tomatometer_rating",
    "IMDB Rating": "imdb_rating",
    "Tomatometer Status": "tomatometer_status"}, inplace= True)

osc_df.info()

<class 'pandas.DataFrame'>
Index: 438 entries, 0 to 561
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   title               438 non-null    str           
 1   oscar_year          438 non-null    str           
 2   studio_producer     438 non-null    str           
 3   oscar_award         438 non-null    str           
 4   runtime             438 non-null    float64       
 5   movie_genre         438 non-null    str           
 6   imdb_rating         438 non-null    float64       
 7   overview            438 non-null    str           
 8   genres              438 non-null    str           
 9   critic_consensus    338 non-null    str           
 10  content_rating      438 non-null    str           
 11  directors           436 non-null    str           
 12  authors             436 non-null    str           
 13  release_date        438 non-null    datetime64[us]
 14  production

### Normalizing Genre Columns  
**horror_df: movie_genres**  
^ string with comma seperating each genre; how I want each dataframe's genre column to be.
  
bech_df: genre1, genre2, genre3  
osc_df: movie_genre, genres


In [445]:
# create a function to unify the genre columns into one column for each dataframe
# string separated by commas

def unify_genres(df, genre_columns):
    for col in genre_columns:
        df[col] = df[col].fillna('').astype(str).apply(lambda x: x.split(',') if ',' in x else [x] if x else [])
    df['movie_genres'] = df[genre_columns].apply(
        lambda row: list({g.strip() for col in genre_columns for g in row[col] if g.strip()}),
        axis=1
    )
    df['movie_genres'] = df['movie_genres'].apply(lambda x: ', '.join(x))
    return df


# mixture of using chatgpt and www.slingacademy.com
# the code chatgpt gave me didn't work exactly the way I had hoped and needed modifications to get the desired output

unify_genres(bech_df, ['genre1', 'genre2', 'genre3'])
unify_genres(osc_df, ['movie_genre', 'genres'])

,title,oscar_year,studio_producer,oscar_award,runtime,movie_genre,imdb_rating,overview,genres,critic_consensus,content_rating,directors,authors,release_date,production_company,tomatometer_status,tomatometer_rating,movie_genres
0,Wings,1927/28,Famous Players-Lasky,Winner,144.0,"[Drama, Romance, War]",7.5,"With World War I afoot, David Armstrong (Richa...","[Action & Adventure, Classics]",Subsequent war epics may have borrowed heavily...,PG-13,William Wellman,"Hope Loring, Louis D. Lighton",1927-08-12,Unknown,Certified-Fresh,93.0,"Romance, Classics, Drama, War, Action & Adventure"
3,The Broadway Melody,1928/29,Metro-Goldwyn-Mayer,Winner,100.0,"[Drama, Musical, Romance]",5.7,"Vaudeville sisters ""Hank"" (Bessie Love) and Qu...","[Musical & Performing Arts, Romance]",The Broadway Melody is interesting as an examp...,NR,Harry Beaumont,"James Gleason, Norman Houston, Sarah Y. Mason",1929-02-01,MGM Home Entertainment,Rotten,33.0,"Romance, Musical, Drama, Musical & Performing ..."
6,In Old Arizona,1928/29,Fox,Nominee,95.0,[Western],5.6,"In this early Western, notorious bandit the Ci...","[Action & Adventure, Classics, Western, Rom...",NaN,NR,"Irving Cummings, Raoul Walsh",Tom Barry,1928-12-25,20th Century Fox Film Corporation,Rotten,56.0,"Romance, Classics, Western, Action & Adventure"
9,The Big House,1929/30,Cosmopolitan,Nominee,87.0,"[Crime, Drama, Thriller]",7.1,"After being convicted of manslaughter, drunk d...","[Classics, Drama]",NaN,NR,George W. Hill,"Frances Marion, Martin Flavin, Joseph Farnham,...",1930-05-14,Warner Bros.,Fresh,75.0,"Thriller, Crime, Classics, Drama"
13,Cimarron,1930/31,RKO Radio,Winner,123.0,"[Drama, Western]",5.9,"In 1889, adventurous lawyer and newspaper edit...","[Classics, Drama, Western]",Cimarron is supported by a strong performance ...,NR,Wesley Ruggles,"Howard Estabrook, Edna Ferber",1931-01-26,MGM Home Entertainment,Rotten,50.0,"Classics, Drama, Western"
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
552,A Star Is Born,2018,"Bill Gerber, Bradley Cooper, and Lynette Howel...",Nominee,136.0,"[Drama, Music, Romance]",7.6,Seasoned musician Jackson Maine discovers -- a...,[Drama],"With appealing leads, deft direction, and an a...",R,Bradley Cooper,"Will Fetters, Eric Roth, Bradley Cooper",2018-10-05,Warner Bros. Pictures,Certified-Fresh,90.0,"Romance, Music, Drama"
553,Vice,2018,"Dede Gardner, Jeremy Kleiner, Adam McKay, and ...",Nominee,132.0,"[Biography, Comdey, Drama]",7.2,Gov. George W. Bush of Texas picks Dick Cheney...,[Drama],"Vice takes scattershot aim at its targets, but...",R,Adam McKay,Adam McKay,2018-12-25,Annapurna Pictures,Fresh,65.0,"Biography, Comdey, Drama"
557,The Irishman,2019,"Martin Scorsese, Robert De Niro, Jane Rosentha...",Nominee,209.0,"[Biography, Crime, Drama]",7.8,"In the 1950s, truck driver Frank Sheeran gets ...","[Drama, Mystery & Suspense]",An epic gangster drama that earns its extended...,R,Martin Scorsese,Steven Zaillian,2019-11-01,Netflix,Certified-Fresh,96.0,"Mystery & Suspense, Crime, Biography, Drama"
559,Joker,2019,"Todd Phillips, Bradley Cooper, and Emma Tillin...",Nominee,122.0,"[Crime, Drama, Thriller]",8.4,"Forever alone in a crowd, failed comedian Arth...","[Action & Adventure, Drama, Mystery & Suspense]",Joker gives its infamous central character a c...,R,Todd Phillips,"Todd Phillips, Scott Silver",2019-10-04,Warner Bros. Pictures,Fresh,68.0,"Crime, Thriller, Drama, Mystery & Suspense, Ac..."


Now I can drop the columns that have been unified

In [446]:
osc_df.drop(['movie_genre', 'genres'], axis=1, inplace=True)
bech_df.drop(['genre1', 'genre2', 'genre3'], axis=1, inplace=True)

## Creating DataFrames  
(to be turned into tables)


movie table, ratings/scores/ awards, financials, cinatomy/subjective, studio/directors/producers
  
  genres, genre id + name --- how to split the comma separated string into indiviual values ? or just add to movie table.

In [447]:
df1= horror_df[['title', 'original_title', 'original_language', 'overview', 'tagline', 'collection_name', 'runtime', 'release_date', 'movie_genres']]
df2= osc_df[['title', 'overview', 'runtime', 'release_date', 'movie_genres']]
df3= bech_df[['title','runtime', 'release_date', 'movie_genres']]
df4= cin_df[['title']]



movies_info = pd.concat([df1, df2, df3, df4], ignore_index=True, sort=False)

movies_df = movies_info.drop_duplicates().reset_index(drop=True)
movies_df.insert(0, 'movie_id', range(1, len(movies_df) + 1))
movies_df.info()




# checking all movie_id values are unique 
movies_df['movie_id'].nunique()


<class 'pandas.DataFrame'>
RangeIndex: 42290 entries, 0 to 42289
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   movie_id           42290 non-null  int64         
 1   title              42290 non-null  str           
 2   original_title     31142 non-null  str           
 3   original_language  31142 non-null  str           
 4   overview           31580 non-null  str           
 5   tagline            12628 non-null  str           
 6   collection_name    2281 non-null   str           
 7   runtime            39631 non-null  float64       
 8   release_date       41291 non-null  datetime64[ns]
 9   movie_genres       41291 non-null  str           
dtypes: datetime64[ns](1), float64(1), int64(1), str(7)
memory usage: 3.2 MB


42290

In [448]:
# financials table ---  movie_id, budget, revenue 
df1 = movies_df[['movie_id', 'title']]
df2 = horror_df[['title', 'budget', 'revenue']]

fin_info = pd.merge(df1, df2, on="title", how= 'right')
financials_df = fin_info.drop_duplicates().reset_index(drop=True)

#movie_id column datatype needs to be turned back into an integer
financials_df['movie_id'] = financials_df['movie_id'].astype('Int64')


# drop rows with NaN for budget and revenue?
financials_df.info()


<class 'pandas.DataFrame'>
RangeIndex: 35248 entries, 0 to 35247
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   movie_id  35248 non-null  Int64
 1   title     35248 non-null  str  
 2   budget    35248 non-null  int64
 3   revenue   35248 non-null  int64
dtypes: Int64(1), int64(2), str(1)
memory usage: 1.1 MB


In [449]:
# critique table  movie_id, tmdb rating, imdb rating, tomatometer rating, bechdel rating, content rating, oscar award

df1 = movies_df[['movie_id', 'title']]
df2 = horror_df[['title', 'tmdb_rating']]
df3 =osc_df[['title', 'oscar_award', 'oscar_year', 'imdb_rating', 'critic_consensus', 'tomatometer_status', 'tomatometer_rating']]
df4 = bech_df[['title', 'bechdel_rating', 'imdb_rating']]



ratings1 = pd.merge(df2, df3, on='title', how='outer')
ratings2 = pd.merge(ratings1, df4, on= 'title', how= 'outer')

ratings_info = pd.merge(df1, ratings2, on= 'title', how= "right")

ratings_df = ratings_info.drop_duplicates().reset_index(drop=True)


#movie_id column datatype needs to be turned back into an integer
ratings_df['movie_id'] = ratings_df['movie_id'].astype('Int64')

ratings_df.info()



<class 'pandas.DataFrame'>
RangeIndex: 51284 entries, 0 to 51283
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   movie_id            51284 non-null  Int64  
 1   title               51284 non-null  str    
 2   tmdb_rating         41269 non-null  float64
 3   oscar_award         1235 non-null   str    
 4   oscar_year          1235 non-null   str    
 5   imdb_rating_x       1235 non-null   float64
 6   critic_consensus    937 non-null    str    
 7   tomatometer_status  1235 non-null   str    
 8   tomatometer_rating  1235 non-null   float64
 9   bechdel_rating      15447 non-null  float64
 10  imdb_rating_y       15447 non-null  float64
dtypes: Int64(1), float64(5), str(5)
memory usage: 4.4 MB


In [450]:
# affiliates table affiliate_id, stuio, director, producer 
#add affiliate id to movies_df ??? 
df1= movies_df[['movie_id', 'title']]
df2 = osc_df[['title', 'studio_producer', 'directors', 'authors', 'production_company']]

aff_info = pd.merge(df1, df2, on='title', how='right')


affiliates_df = aff_info.drop_duplicates().reset_index(drop=True)

#movie_id column datatype needs to be turned back into an integer
affiliates_df['movie_id'] = affiliates_df['movie_id'].astype('Int64')


affiliates_df.info()

# drop rows where all are NaN except id


<class 'pandas.DataFrame'>
RangeIndex: 963 entries, 0 to 962
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   movie_id            963 non-null    Int64
 1   title               963 non-null    str  
 2   studio_producer     963 non-null    str  
 3   directors           959 non-null    str  
 4   authors             955 non-null    str  
 5   production_company  963 non-null    str  
dtypes: Int64(1), str(5)
memory usage: 46.2 KB


In [451]:
# cinatomy/subjective --- table movie_id, every column i think i'll use except title from cinatomy

df1= movies_df[['movie_id', 'title']]
df2 = cin_df[['title', 'originality', 'immersive', 'impactful', 'connected', 'visuals', 'visual_effects', 'performance_of_actors', 'trailer_or_spoiler', 'plot_quality_rating', 'end_feeling_ending_rating', 'pacing_efficiency' ]]


cin_info = pd.merge(df1, df2, on='title', how='right')
cinatomy_df = cin_info.drop_duplicates().reset_index(drop=True)

#movie_id column datatype needs to be turned back into an integer
cinatomy_df['movie_id'] = cinatomy_df['movie_id'].astype('Int64')


cinatomy_df.info()


<class 'pandas.DataFrame'>
RangeIndex: 1865 entries, 0 to 1864
Data columns (total 13 columns):
 #   Column                     Non-Null Count  Dtype
---  ------                     --------------  -----
 0   movie_id                   1865 non-null   Int64
 1   title                      1865 non-null   str  
 2   originality                1865 non-null   int64
 3   immersive                  1865 non-null   int64
 4   impactful                  1865 non-null   int64
 5   connected                  1865 non-null   int64
 6   visuals                    1865 non-null   int64
 7   visual_effects             1865 non-null   int64
 8   performance_of_actors      1865 non-null   int64
 9   trailer_or_spoiler         1865 non-null   int64
 10  plot_quality_rating        1865 non-null   int64
 11  end_feeling_ending_rating  1865 non-null   int64
 12  pacing_efficiency          1865 non-null   int64
dtypes: Int64(1), int64(11), str(1)
memory usage: 191.4 KB


### drop title column from each dataframe


In [452]:
financials_df.drop(["title"], axis=1, inplace=True)
affiliates_df.drop(["title"], axis=1, inplace=True)
ratings_df.drop(["title"], axis=1, inplace=True)
cinatomy_df.drop(["title"], axis=1, inplace=True)


### Turning DataFrames into CSV Files


In [453]:
movies_df.to_csv('../Data/movies.csv')
financials_df.to_csv('../Data/financials.csv')
affiliates_df.to_csv('../Data/affiliates.csv')
ratings_df.to_csv('../Data/ratings.csv')
cinatomy_df.to_csv('../Data/cinatomy.csv')

### Creating a Database and load new dataframes as tables

In [454]:

connection = sqlite3.connect('../Data/perfect_movie.db')

movies_df.to_sql('movies', connection, if_exists='replace', index=False)
financials_df.to_sql('financials', connection, if_exists='replace', index=False)
affiliates_df.to_sql('affiliates', connection,  if_exists='replace',index=False)
ratings_df.to_sql('ratings', connection, if_exists='replace', index=False)
cinatomy_df.to_sql('cinatomy', connection, if_exists='replace', index=False)


connection.close()

## SQL queries

In [455]:

connection = sqlite3.connect('../Data/perfect_movie.db')

pd.read_sql("SELECT name FROM sqlite_master WHERE type='table';", connection)

,name
0,movies
1,financials
2,affiliates
3,ratings
4,cinatomy


In [456]:
query1 = """ 
SELECT 
FROM
;
"""
pd.read_sql(query1, connection)

DatabaseError: Execution failed on sql ' 
SELECT 
FROM
;
': near "FROM": syntax error